src/mumble/pluginAPI/Plugin.h

```diff
@@ -0,0 +1,139 @@
+	///
+	/// @returns A pointer to the description (encoded as a C-String)
+	PLUGIN_EXPORT const char* getDescription();
+
+	/// Provides the MumbleAPI struct to the plugin. This struct contains function pointers that can be used
+	/// to interact with the Mumble client. It is up to the plugin to store this struct somewhere if it wants to make use
+	/// of it at some point.
+	///
+	/// @param api The MumbleAPI struct
+	PLUGIN_EXPORT void registerAPIFunctions(const struct MumbleAPI *api);
+
+	/// Registers the ID of this plugin. This is the ID Mumble will reference this plugin with and by which this plugin
+	/// can identify itself when communicating with Mumble.
+	///
+	/// @param id The ID for this plugin
+	PLUGIN_EXPORT void registerPluginID(uint32_t id);
+
+
+	// Parameters to functions below are yet to be determined
+
+	// -------- Callback functions -----------
+
+	/// Called when connecting to a server.
+	PLUGIN_EXPORT void onServerConnected();
+
+	/// Called when disconnecting from a server.
+	PLUGIN_EXPORT void onServerDisconnected();
+
+	/// Called whenever any user on the server enters a channel
+	/// This function will also be called when freshly connecting to a server as each user on that
+	/// server needs to be "added" to the respective channel as far as the local client is concerned.
+	///
+	/// @param userID The ID of the user this event has been triggered for
+	/// @param previousChannelID The ID of the chanel the user is coming from. Negative IDs indicate that there is no previous channel (e.g. the user
+	/// 	freshly connected to the server) or the channel isn't available because of any other reason.
+	/// @param newChannelID The ID of the channel the user has entered. If the ID is negative, the new channel could not be retrieved. This means
+	/// 	that the ID is invalid.
+	PLUGIN_EXPORT void onChannelEntered(MumbleUserID_t userID, MumbleChannelID_t previousChannelID, MumbleChannelID_t newChannelID);
+
+	/// Called whenever a user leaves a channel.
+	/// This function will also be called when the user disconnects from the server.
+	///
+	/// @param userID The ID of the user that left the channel
+	/// @param channelID The ID of the channel the user left. If the ID is negative, the channel could not be retrieved. This means that the ID is
+	/// 	invalid.
+	PLUGIN_EXPORT void onChannelExited(MumbleUserID_t userID, MumbleChannelID_t channelID);
+
+	/// Called when any user changes his/her talking state.
+	///
+	/// @param userID The ID of the user whose talking state has been changed
+	/// @param talkingState The new TalkingState the user has switched to.
+	PLUGIN_EXPORT void onUserTalkingStateChanged(MumbleUserID_t userID, TalkingState_t talkingState);
+
+	/// Called whenever there is audio input
+	PLUGIN_EXPORT void onAudioInput(short *inputPCM, bool isSpeech);
+
+	/// Called whenever there is audio output
+	PLUGIN_EXPORT void onAudioOutput_short(short *outputPCM, uint32_t sampleCount, uint32_t channelCount);
```

Comment:
just for consistency. Why use short and not int16_t? The rest of API function arguments use fixed length integers.

Reply:
Mumble uses `short` internally. And although a `short` is _at least` 16 bit, it might be longer (to my understanding) so forcing this to be an `int16_t` could cause trouble in such a case.

Reply:
You could convert it internally.

Reply:
What good would that do? Is knowing the exact size of the used data type when writing the plugin that important?
To me this seems a little like wasting processing time...

Reply:
Fixed size integers increase robustness of any code that has to be cross-platform. Now you may end having a system with short as 16 bits and another as 24 bits. In the first you may have overflow problems, in the second you won't be able to reproduce as you may not reach MAX_INT16. At the end it's just robustness of your system.

Reply:
Yeah that makes sense. I'll have to look into how such a conversion could be done - I'm not quite sure how this will work yet.

However if this isn't super complicated or has a huge performance impact I think we can use fixed size integers :+1:

Reply:
Fixed size integers have no impact on performance. There are even _fast_ variants

Reply:
I was more referring to the conversion not the types themselves

Reply:
Alright I googled around quite a bit but I can't seem to find a way that would let be convert a `short` to `int16_t ` if the size of the `short` is bigger than the size of our `int16_t`.
The only solution I could come up with for this problem is to use a `int32_t` of which I'd hope to be of bigger size than a `short`in every implementation...

Reply:
You would have then the same problem but inverse if you use int32_t. Even right now on windows because short is 2 bytes singoes int. A conversion check to see if overflows and throw an error is as simple as comparing it to MAX_INT16 and MIN_}INT16 (or similar names). Both are defined in stdint.h

Reply:
You would have then the same problem but inverse if you use int32_t. Even right now on windows because short is 2 bytes singoes int. A conversion check to see if overflows and throw an error is as simple as comparing it to MAX_INT16 and MIN_}INT16 (or similar names). Both are defined in stdint.h

Reply:
Ah so all you'd do is to throw an error if the transformation doesn't work?
Seems a bit weird to do it this way... On the other hand there isn't really a way to do it differently, I guess :thinking:

Alright then. I'll replace it with `int16_t`